contracts/src/registry/L1ETHRegistry.sol

```diff
@@ -0,0 +1,233 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.13;
+
+import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
+
+import {IL1EjectionController} from "../controller/IL1EjectionController.sol";
+import {ERC1155Singleton} from "./ERC1155Singleton.sol";
+import {IERC1155Singleton} from "./IERC1155Singleton.sol";
+import {IRegistry} from "./IRegistry.sol";
+import {IRegistryDatastore} from "./IRegistryDatastore.sol";
+import {BaseRegistry} from "./BaseRegistry.sol";
+import {PermissionedRegistry} from "./PermissionedRegistry.sol";
+
+/**
+ * @title L1ETHRegistry
+ * @dev L1 contract for .eth that holds ejected .eth names only.
+ * Unlike the L2 ETHRegistry, this registry does not handle new registrations directly,
+ * but receives names that have been ejected from L2.
+ */
+contract L1ETHRegistry is PermissionedRegistry, AccessControl {
+    bytes32 public constant EJECTION_CONTROLLER_ROLE = keccak256("EJECTION_CONTROLLER_ROLE");
+
+    error NameExpired(uint256 tokenId);
+    error NameNotExpired(uint256 tokenId, uint64 expires);
+    error CannotReduceExpiration(uint64 oldExpiration, uint64 newExpiration);
+
+    event NameEjected(uint256 indexed tokenId, address owner, uint64 expires);
+    event NameRenewed(uint256 indexed tokenId, uint64 newExpiration, address renewedBy);
+    event NameMigratedToL2(uint256 indexed tokenId, address sendTo);
+    event EjectionControllerChanged(address oldController, address newController);
+
+    address public ejectionController;
+
+    constructor(IRegistryDatastore _datastore, address _ejectionController) PermissionedRegistry(_datastore) {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+
+        // Grant role to the ejection controller
+        require(_ejectionController != address(0), "Ejection controller cannot be empty");
+
+        ejectionController = _ejectionController;
+        _grantRole(EJECTION_CONTROLLER_ROLE, _ejectionController);
+    }
+
+    /**
+     * @dev Set a new ejection controller
+     * @param _newEjectionController The address of the new controller
+     */
+    function setEjectionController(address _newEjectionController) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(_newEjectionController != address(0), "Ejection controller cannot be empty");
+        
+        address oldController = ejectionController;
+        
+        // Revoke role from the old controller
+        _revokeRole(EJECTION_CONTROLLER_ROLE, oldController);
+        
+        // Set the new controller and grant role
+        ejectionController = _newEjectionController;
+        _grantRole(EJECTION_CONTROLLER_ROLE, _newEjectionController);
+        
+        emit EjectionControllerChanged(oldController, _newEjectionController);
+    }
```

Comment:
Given that there's only one ejection controller at a time, and we track it in a variable, we can eliminate the unnecessary use of the roles modifier here, and just directly check if the caller is the current ejection controller.